import os
import yaml

from launch import LaunchDescription
from launch.actions import (
    DeclareLaunchArgument,
    IncludeLaunchDescription,
    OpaqueFunction,
)
from launch.conditions import IfCondition
from launch.launch_description_sources import PythonLaunchDescriptionSource
from launch.substitutions import Command, FindExecutable, PathJoinSubstitution, LaunchConfiguration
from launch_ros.actions import Node
from launch_ros.substitutions import FindPackageShare

from ament_index_python.packages import get_package_share_directory

_LAUNCHARG_KEY_DEBUGVIZ = "debug_visual"
_LAUNCHARG_KEY_RVIZCONF = "rviz_conf"

def load_file(package_name, file_path):
    package_path = get_package_share_directory(package_name)
    absolute_file_path = os.path.join(package_path, file_path)

    try:
        with open(absolute_file_path, "r") as file:
            return file.read()
    except EnvironmentError:  # parent of IOError, OSError *and* WindowsError where available
        return None

def load_yaml(package_name, file_path):
    package_path = get_package_share_directory(package_name)
    absolute_file_path = os.path.join(package_path, file_path)

    try:
        with open(absolute_file_path, "r") as file:
            return yaml.safe_load(file)
    except EnvironmentError:  # parent of IOError, OSError *and* WindowsError where available
        return None

def launch_setup(context, *args, **kwargs):
    debug_visual = LaunchConfiguration(_LAUNCHARG_KEY_DEBUGVIZ)
    rviz_config_file = LaunchConfiguration(_LAUNCHARG_KEY_RVIZCONF)

    # Generate Robot Description parameter from xacro
    robot_description_content = Command(
        [
            PathJoinSubstitution([FindExecutable(name="xacro")]),
            " ",
            PathJoinSubstitution([FindPackageShare("ariac_description"), "urdf/ariac_robots", "ariac_robots.urdf.xacro"]), 
            " ",
        ]
    )
    robot_description = {"robot_description": robot_description_content}

    ## Moveit Parameters
    robot_description_semantic = {"robot_description_semantic": load_file("ariac_moveit_config", "srdf/ariac_robots.srdf")}

    robot_description_kinematics = {"robot_description_kinematics": load_yaml("ariac_moveit_config", "config/kinematics.yaml")}

    ompl_planning_pipeline_config = {
        "move_group": {
            "planning_plugin": "ompl_interface/OMPLPlanner",
            "request_adapters": """default_planner_request_adapters/AddTimeOptimalParameterization default_planner_request_adapters/FixWorkspaceBounds default_planner_request_adapters/FixStartStateBounds default_planner_request_adapters/FixStartStateCollision default_planner_request_adapters/FixStartStatePathConstraints""",
            "start_state_max_bounds_error": 0.1,
        }
    }
    ompl_planning_yaml = load_yaml("ariac_moveit_config", "config/ompl_planning.yaml")

    ompl_planning_pipeline_config["move_group"].update(ompl_planning_yaml)
    
    moveit_controllers = {
        "moveit_simple_controller_manager": load_yaml("ariac_moveit_config", "config/controllers.yaml"),
        "moveit_controller_manager": "moveit_simple_controller_manager/MoveItSimpleControllerManager",
    }

    trajectory_execution = {
        "moveit_manage_controllers": False,
        "trajectory_execution.allowed_execution_duration_scaling": 1.2,
        "trajectory_execution.allowed_goal_duration_margin": 0.5,
        "trajectory_execution.allowed_start_tolerance": 0.01,
    }

    planning_scene_monitor_parameters = {
        "publish_planning_scene": True,
        "publish_geometry_updates": True,
        "publish_state_updates": True,
        "publish_transforms_updates": True,
    }

    # Move group node
    move_group_node = Node(
        package="moveit_ros_move_group",
        executable="move_group",
        output="screen",
        parameters=[
            robot_description,
            robot_description_semantic,
            robot_description_kinematics,
            ompl_planning_pipeline_config,
            trajectory_execution,
            moveit_controllers,
            planning_scene_monitor_parameters,
            {"use_sim_time": True},
        ],
    )
  
    rviz_node = Node(
        package="rviz2",
        executable="rviz2",
        name="rviz2_moveit",
        output="log",
        arguments=["-d", rviz_config_file],
        parameters=[
            robot_description,
            robot_description_semantic,
            robot_description_kinematics,
            {"use_sim_time": True}
        ],
        condition=IfCondition(debug_visual)
    )
    
    nodes_to_start = [
        move_group_node,
    ]
    if rviz_node: nodes_to_start.append(rviz_node)

    return nodes_to_start


def generate_launch_description():
    # rviz with moveit configuration
    _launcharg_rviz_config = DeclareLaunchArgument(
        _LAUNCHARG_KEY_RVIZCONF,
        default_value= PathJoinSubstitution(
            [FindPackageShare("ariac_moveit_config"), "rviz", "moveit.rviz"]))
    _launcharg_mode_debugviz = DeclareLaunchArgument(
        _LAUNCHARG_KEY_DEBUGVIZ,
        default_value="false"
        )
    
    declared_arguments = [_launcharg_rviz_config, _launcharg_mode_debugviz]

    return LaunchDescription(declared_arguments + [OpaqueFunction(function=launch_setup)])
